Ping CX tracker every 5 min

sn-srdjan opened the pull request:
Upon CX application start ping tracker service with current config.

Commit:
Ping CX tracker every 5 min

Upon CX application start ping tracker service with current config.

Commit:
Replace cx tracker cli with import

Instead of relying on built cx-tracker-cli use dep ensure

Commit:
Update Tracker CLI dependency

Commit:
Use configured values to start cx chain

Add needed flags for config to be used
Use configured values as primary source (consider switching)
Add suport for config download based on its hash value
Lower the level for ping tracker issue

sn-srdjan commented on cxgo/main.go:
```diff
@@ -319,6 +322,35 @@ func main() {
+			panic(errors.New(fmt.Sprintln("config from path %v can't be unmarshalled to JSON struct due to error %v", options.configPath, err)))
+		}
+
+		// TODO we probably need to do few more steps here like creating the wallet and
```
Here are two points to be discussed - should wallet be created automatically and those configs updated.

As well as second part, there is no point in running with --blockchain flag as far as I understand this topic.

sn-srdjan replied:
Additional thing to discuss is - should potential command line arguments overwrite config value?
Right now config has highest priority and I'm not sure is it even a valid use case to consider mixing those two?

amherag replied:
Automatic creation of wallets could be good. Wallets don't really make sense for CX chains, as we're managing programs, not coins. Wallets are used to sign transactions, but I don't see any reason to have several wallets.

You need to run with `--blockchain` if you need to generate a genesis signature.

amherag replied:
Most projects I've used usually prioritize command-line arguments. And, at least to me, this makes sense: if a user is explicitly adding a command line argument at the moment of wanting to run your application, then it means that the user really wants that configuration setting to be the one to be used. Environment variables should be the default, then config files should override these, and then command-line arguments should override everything.

sn-srdjan commented on cxgo/main.go:
```diff
@@ -352,6 +384,24 @@ func main() {
+			log.Info("Pinging CX Tracker is disabled")
+		} else {
+			log.Info("Initiating up CX Tracker ping")
+			pingTracker(options.configPath) // TODO should we ping on startup or after first period of 5 min is over and we're sure instance is running correctly?
```
Not sure if we discussed this. Think we said it's ok to ping immediately but just to double check that statement.

amherag replied:
I think it's okay.

sn-srdjan commented on vendor/github.com/sn-srdjan/cx-tracker-cli/src/cli/cxtracker/tracker_provider.go:
```diff
@@ -0,0 +1,91 @@
+
+// CXApplicationConfig holds config describing CX Application instance
+type CXApplicationConfig struct {
+	//TODO confirm should config contain genesis hash and work based on it (alternative needing to use config file content hash), and
```
This is another important topic.
One part is trivial - to confirm is genesis hash the same thing as genesis signature. I saw both referenced in the project, think they are the same but would like to double check that info. Signature is mentioned in steps needed to setup the chain and I know hash was mentioned before and I also se references in the code so...

Another is not so trivial. We talked about having CX Tracker store CX Application configs.
There is a concept of hash representing content of the config file. And there is concept of Genesis Hash (if previous discussion topic makes any sense).
Pros and Cons of using either of two hashes is detailed in code so let's discuss on how to proceed with this.

amherag replied:
First point: Yes, genesis hash is the same as the genesis signature.

Second point: As far as I understand, we only need a way to uniquely identify the CX application, i.e. we need anything that will be unique and that it will be attached to that CX application. If you already have access to the genesis hash, and this will be unique, I think it can be used to access information for that entry (like retrieving the config file).

sn-srdjan commented on vendor/github.com/sn-srdjan/cx-tracker-cli/src/cli/cxtracker/tracker_provider.go:
```diff
@@ -0,0 +1,91 @@
+
+	f, err := os.Create(configFilePath)
+	if err != nil {
+		return fmt.Errorf("error while creating file for config with genesis hash: %s due to error: %s", genesisHash, err)
```
Should discuss a couple of scenarios here:
- if we don't have rights or parts of the path don't exist we should just fail
- if there is already config file on given path right now it's overwritten - not sure if we should do something about that (keep old as backup or something else)

amherag replied:
* I don't know how we could not have rights to write to the path. Are you trying to write somewhere in root or somewhere obscure? Then it will be a problem, but this shouldn't be the case. Everything should be written to the user directory or somewhere where we know for sure the user has rights to write to. In case the user did something really weird (which happens more than I ever imagined), yeah, sure, throw an error.

* I'm not sure if I have the whole picture in my mind about this, but I see that `options.configPath` represents a **filename**, which means that the user has the capability of specifying multiple configuration files, each for a different CX chain. If the user decides to create another CX chain with the same config filename, we can assume that the user wants to overwrite any existing config files. It'd be extremely annoying to be testing CX chains and find out later on that you have dozens of backup files of config files that you'll never use again.

Commit:
Replace cx tracker CLI source repo

Commit:
Update startup procedure

Add warn message when provided flag is overwritten by config
Remove old TODOs that are confirmed

Still to do:
Make sure everything is working fine when replacing cx apps locally

Commit:
Added example and tracker readme

Commit:
Improve CX app run

- Auto generate wallet in case one is missing on environment.
- Check last run app genesis hash.
Reset local database if we're running different app now.
